lib/value/src/kind/accessor.rs

```diff
@@ -0,0 +1,543 @@
+                        "foo".into(),
+                        Kind::array(BTreeMap::from([
+                            (1.into(), Kind::integer()),
+                            (
+                                2.into(),
+                                Kind::object(BTreeMap::from([
+                                    (
+                                        "bar".into(),
+                                        Kind::object(BTreeMap::from([(
+                                            "baz".into(),
+                                            Kind::integer().or_regex(),
+                                        )])),
+                                    ),
+                                    ("qux".into(), Kind::boolean()),
+                                ])),
+                            ),
+                        ])),
+                    )])),
+                    path: LookupBuf::from_str(".foo[2].bar").unwrap(),
+                    want: Some(Kind::object(BTreeMap::from([(
+                        "baz".into(),
+                        Kind::integer().or_regex(),
+                    )]))),
+                },
+            ),
+        ]) {
+            assert_eq!(
+                kind.find_at_path(&path.to_lookup()),
+                want.as_ref(),
+                "returned: {}",
+                title
+            );
+        }
+    }
+
+    #[test]
+    fn test_nest_at_path() {
+        struct TestCase {
+            kind: Kind,
+            path: LookupBuf,
+            want: Kind,
+        }
+
+        for (title, TestCase { kind, path, want }) in HashMap::from([
+            (
+                "single-level object",
+                TestCase {
+                    kind: Kind::bytes(),
+                    path: "foo".into(),
+                    want: Kind::empty().or_object(BTreeMap::from([("foo".into(), Kind::bytes())])),
+                },
+            ),
+            (
+                "multi-level object",
+                TestCase {
+                    kind: Kind::boolean(),
+                    path: LookupBuf::from_str("foo.bar").unwrap(),
+                    want: Kind::empty().or_object(BTreeMap::from([(
+                        "foo".into(),
+                        Kind::object(BTreeMap::from([("bar".into(), Kind::boolean())])),
+                    )])),
+                },
+            ),
+            (
+                "array positive index",
+                TestCase {
+                    kind: Kind::integer(),
+                    path: LookupBuf::from_str("[2]").unwrap(),
+                    want: Kind::empty().or_array(BTreeMap::from([(2.into(), Kind::integer())])),
+                },
+            ),
+            (
+                "array negative index",
+                TestCase {
+                    kind: Kind::integer(),
+                    path: LookupBuf::from_str("[-2]").unwrap(),
+                    want: Kind::empty().or_array(BTreeMap::default()),
+                },
```

Comment:
Yeah, this behaviour is questionable. I'll think on this, and will circle back to this in the follow-up PR that links it up to VRL, if I think there's a better way to handle this (preferably without panicking, and most-likely also without returning an error because there are places in VRL where this method is called that doesn't allow returning an error. Although I might opt to return an error for this method, but then panic at the VRL call-site, if we deem it to be a programming error).